chaos-days/blog/2024-01-19-Job-Activation-Latency/index.md

```diff
@@ -0,0 +1,222 @@
+categories: 
+  - performance 
+  - bpmn
+tags:
+  - availability
+authors: nicolas
+---
+
+# Reducing the job activation delay
+
+With the addition of end-to-end job streaming capabilities in Zeebe, we wanted to measure the improvements in job activation latency:
+
+- How much is a single job activation latency reduced?
+- How much is the activation latency reduced between each task of the same process instance?
+- How much is the activation latency reduced on large clusters with a high broker and partition count?
+
+Additionally, we wanted to guarantee that every component involved in streaming, including clients, would remain resilient in the face of load surges.
+
+**TL;DR;** Job activation latency is greatly reduced, with task based workloads seeing up to 50% reduced overall execution latency. Completing a task now immediately triggers pushing out the next one, meaning the latency to activate the next task in a sequence is bounded by how much time it takes to process its completion in Zeebe. Activation latency is unaffected by how many partitions or brokers there in a cluster, as opposed to job polling, thus ensuring scalability of the system. Finally, reuse of gRPC's flow control mechanism ensure clients cannot be overloaded even in the face of load surges, without impacting other workloads in the cluster.
+
+[Head over to the documentation to learn how to start using job push!](https://docs.camunda.io/docs/components/concepts/job-workers/#job-streaming)
+
+<!--truncate-->
+
+## Why job activation latency matters
+
+Jobs are one of the fundamental building blocks of Zeebe, representing primarily all tasks (e.g. service, send, user), as well as some less obvious symbols (e.g. intermediate message throw event). In essence, they represent the actual unit of work in a process, the part users will implement, i.e. the actual application code. To reduce the likelihood of a job being worked on by multiple clients at the same time, it first goes through an activation process, where it is soft-locked for a specific amount of time. Soft-locked here means anyone can still interact with it - they can complete the job, fail it, etc. Only the activation is locked out, meaning no one else can activate the job until it's timed out.
+
+This means that most workloads will consist mostly of job interactions: creation, activation, completion, etc. As such, it's critical to ensure clients receive jobs as fast as possible in order to make progress.
+
+
+## Polling: a first implementation
+
+Back in 2018, Zeebe introduced the `ActivateJobs` RPC for its gRPC clients, analogous to fetching and locking [external tasks in Camunda 7.x](https://docs.camunda.org/manual/7.20/user-guide/process-engine/external-tasks/). This endpoint allowed clients to activate fetch and activate a specific number of available jobs. In other words, it allowed them to _poll_ for jobs.
+
+This was the first implementation to activate and work on jobs in Zeebe for multiple reason:
```

Comment:
I'm wondering whether this is true we had topic subscriptions but not sure whether we had jobs here.  But I guess this doesn't matter 😄